### PR TITLE
Block selection changes

### PR DESCRIPTION
**Pull request by pickedusername11**
No block is selected by default.
Added keys to select previous/next block - this makes it easier to play with one hand.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,4 +34,5 @@ CONTROLS
 - **1/2/3** TO CHOOSE BLOCK
 - **ARROW KEYS** TO MOVE BLOCK
 - **SPACE/RETURN** TO PLACE BLOCK
-- **ESCAPE/Q** TO QUIT
```
Comment (trosh) [inline README.md:37]: why git thinks this was modified still baffles me

File: main.c [hunk 2]
```diff
@@ -31,7 +30,8 @@ int main(int argc, char **argv) {
     char c;      /* INPUT CHARACTER / BLOCK NUMBER */
     char bks[3]; /* BLOCKS GIVEN TO USER */
     char bka[3]; /* FLAGS OF BLOCKS AVAILABLE ; TODO -> BITARRAY */
-    char b;      /* INDEX OF USER CHOSEN BLOCK */
+    char b;      /* INDEX OF USER CHOSEN BLOCK (3 = NO SELECTION) */
+    char oldb;   /* PREVIOUSY SELECTED BLOCK, REQUIRED FOR NEXT/PREV BLOCK */
```
Comment (trosh) [inline main.c:34]: well, only required if you don't want the position to reset when switching blocks with no other block available

File: main.c [hunk 3]
```diff
@@ -95,7 +97,7 @@ int main(int argc, char **argv) {
             if (bka[0]) printbk(bks[0],  0, 12, 3);
             if (bka[1]) printbk(bks[1],  5, 12, 4);
             if (bka[2]) printbk(bks[2], 10, 12, 5);
-            printbk(bks[b], x, y, b+3);
+            if (b <3)   printbk(bks[b], x, y, b+3);
```
Comment (trosh) [inline main.c]: heh, `<3` is cute, but maybe use `-1` for no selection, and `b >= 0` for the test (I think `3` seems to much of a legal value)

File: main.c [hunk 4]
```diff
@@ -114,29 +116,49 @@ int main(int argc, char **argv) {
...
                 break;
+            /* SELECT PREVIOUS BLOCK */
+            case 44 : /* , */
+            case 60 : /* < */
+                if (b < 3) { /* IF SOME BLOCK IS SELECTED */
+                    oldb = b;
+                    while (!bka[++b%3]) {}
+                    b %= 3;
+                    if (oldb != b) x = y = 5;
+                }
```
Comment (trosh) [inline main.c:127]:
```c
else b = 0;
```
?

File: main.c [hunk 5]
```diff
@@ -114,29 +116,49 @@ int main(int argc, char **argv) {
                 b = 2;
                 x = y = 5;
                 break;
+            /* SELECT PREVIOUS BLOCK */
+            case 44 : /* , */
+            case 60 : /* < */
+                if (b < 3) { /* IF SOME BLOCK IS SELECTED */
+                    oldb = b;
+                    while (!bka[++b%3]) {}
```
Comment (trosh) [inline main.c:124]:
```c
b += 3;
do b--; while (!bka[b%3]);
```
or
```c
b += 2;
while (!bka[b%3]) b--;
```

File: main.c [hunk 6]
```diff
@@ -114,29 +116,49 @@ int main(int argc, char **argv) {
...
+                    b %= 3;
+                    if (oldb != b) x = y = 5;
+                }
+                break;
+            /* SELECT NEXT BLOCK */
+            case 46 : /* . */
+            case 62 : /* > */
+                if (b < 3) { /* IF SOME BLOCK IS SELECTED */
+                    oldb = b;
+                    while (!bka[++b%3]) {}
```
Comment (trosh) [inline main.c:134]:
```c
do ++b; while (!bka[b%3]);
```

File: main.c [hunk 7]
```diff
@@ -114,29 +116,49 @@ int main(int argc, char **argv) {
...
+                    b %= 3;
+                    if (oldb != b) x = y = 5;
+                }
+                break;
             case 4 : /* LEFT */
-                if (x > 0
+                if (b < 3 && x > 0
```
Comment (trosh) [inline main.c:140]: If this got too repetitive, you could maybe leave these cases in another `switch` executed only if a block is selected.

File: main.c [hunk 8]
```diff
@@ -149,11 +171,19 @@ int main(int argc, char **argv) {
...
-        for (c=0; c<3; c++) { /* CHOOSE NEW B */
-            if (b == c /* DON'T CHECK BKA[B] */
-             || !bka[c]) continue;
-            b = c;
-            break;
+        /* NO DEFAULT SELECTION UNLESS ONLY ONE BLOCK LEFT */
+        b = 0;
+        for (c=0; c<3; c++) { /* COUNT NUMBER OF AVAILABLE BLOCKS*/
+            b += bka[c];
+        }
```
Comment (trosh) [inline main.c:178]: ooookay i just understood why you're doing this (selecting directly when there's only one left).
I feel like it makes sense but at the same time it slightly breaks the flow of choosing your block.

Testing if any block is left could be reduced to `if (bka[0] || bka[1] || bka[2])`
if you wanted the code to look homogeneously obfuscated lol (actually that feels more readable than temporarily using `b` to count available blocks).

**Review by trosh — CHANGES_REQUESTED** [review]
Please change the `<` / `>` stuff however you like, and the rest if you feel like it.

**Comment by trosh** [comment]
This is the first time someone contributed to a project by me so it's great,
thanks for doing this, and I still have no idea how you found this.

You didn't need to re-fork your repo just because our histories diverged.
If you accepted to discard your side of history you could have done something like
```sh
$ git remote add trosh https://github.com/trosh/1010
$ git fetch trosh master
$ git reset --hard trosh/master
```

The grid is quite nice, I'll look into coloring the borders between colored tiles.

**Comment by pickedusername11** [comment]
Your notes look reasonable, apart from the bit about the </> comment - didn't get what's wrong there.
But, this is where my knowledge of github ends. I have no idea how to make/accept changes.
Feel free to incorporate my changes any way you like.

I found you by searching for a console version of 1010 ;o)
I only use the web interface, not the command-line git, this is why I did the most obvious thing to catch up with your changes.

**Comment by trosh** [comment]
`<`/`>` should select blocks when `b == 3` (`<` should set `b = 2` and `>` should set `b = 0`)

Also currently `<` acts like `>` (you increment b in both cases).
I was suggesting two ways of decrementing (while avoiding going below zero which breaks the modulo).

Great to know that you had the same idea for the game :smiley:

I can only suggest using Git, because while Github is nice for communicating,
real magic is always in Git which will always always work.

I was not going to incorporate my changes directly (like I did yesterday), instead this time I'm just giving feedback and you can update your `master`.

File: main.c (hunk 8 above)
Comment (trosh) [reply main.c:178]: oops that should be `if (bka[0] + bka[1] + bka[2] == 1)`

**Comment by pickedusername11** [comment]
Ah, I see where the confusion comes from.
It was not meant to be three keys '<', '/' and '>', that meant either < or >.
More like scrolling left and right. The slash on it's own does nothing.
Similarly, ',' and '.' were supposed to work the same way as '<' and '>", just without the shift key.
One day I may indeed learn a bit of git. Right now I hate it with passion.

**Comment by trosh** [comment]
You're not getting me &mdash; I'm aware of what you're trying to do, I'm just saying right now it's not doing that. `<` and `,` both do like `>` and `.`; they don't select leftwards but righwards.
```c
case 60 : /* < */
    if (b < 3) { /* IF SOME BLOCK IS SELECTED */
        oldb = b;
        while (!bka[++b%3]) {}
```
This increments `b`, just like the following case does. I'm telling you to decrement instead, but warn you to beware of negative modulo (-1 mod 3 is -1, not 2 as one might expect), and I commented two ways to do that.

Also I think that when nothing is selected, it's annoying to press one of these "scroll" keys with nothing happening, so I suggest allowing `b` to go from 3 to 0 or 2 depending on the direction.

Furthermore I'm personally not fond of using these keys instead of 1/2/3, and I don't even know how it's practical to you to use only one hand &mdash; but since you added it it's fine by me.

Git is love, Git is life

**Comment by pickedusername11** [comment]
Ah, OK. Now I get it.
These were not supposed to be used instead of 1,2,3. Just in addition to.
To have both absolute and relative block changing.

**Comment by pickedusername11** [comment]
I just thought a bit, and think you are right. To select an initial block you have to use 1,2,3 anyway. So no easy one-handed playing. Just throw away that < / > bits then.

Regarding the grid - do you think it may make sense to use the underline attribute for horizontal grid lines? That would probably save some vertical space.

**Comment by trosh** [comment]
OK I integrated your changes, and I kept the `<`/`>` stuff, see if you like how it acts now. Otherwise you can remove that bit of code.

I also improved the display of blocks (connected the tiles together).

I'm thinking that using Unicode upper and left half blocks and corner blocks (or a curses equivalent if there's such a thing) alongside ACS lines could allow filling the grid tiles when coloring. See what I mean ? the grid wouldn't be visible inside blocks (or filled areas) though, so it might be better to use the existing display in the hint board and a filled display in the game board (you can visualize the grid with the surrounding area).

**Comment by pickedusername11** [comment]
Looks great. Thank you.
One minor suggestion - change the clean target in the Makefile to:

clean:
        rm -f ./1010

I think this is a better way to avoid the error message if ./1010 does not exist.
Regarding the upper/lower half blocks - check this link:
http://owlwd.tk/2014/01/30/displaying-conway-s-game-of-life-using-ncurses
Do you like any of the approaches tried there?

**Comment by trosh** [comment]
The article is a good example of using the block characters (in a possibly portable way, not sure), but it doesn't give any more answer for getting a visual grid to look good.

**Comment by trosh** [comment]
hmm my idea of using unicode corner blocks is doomed to fail because some situations would require three colors in a cell

**Pull request closed without merging**